R/symptome_compartment_class.R

```diff
@@ -0,0 +1,226 @@
+                errors <- c(errors, msg)
+              }
+            }
+            # check whether the lengths of nu and pdeath correspond to the number of infected groups
+            n_nu <- length(object@parameters$nu)
+            if (n_nu != 2) {
+              msg <- paste0(
+                "Length of parameter nu,", n_nu, ", is not equal to the setting ", 2)
+              errors <- c(errors, msg)
+            }
+            n_pdeath <- length(object@parameters$pdeath)
+            if (n_pdeath != 2) {
+              msg <- paste0(
+                "Length of parameter pdeath,", n_pdeath, ", is not equal to the setting ", 2)
+              errors <- c(errors, msg)
+            }
+            if (length(errors) == 0) TRUE else errors
+          })
+
+
+#' @describeIn model_symptome to solve the ode system.
+#'
+#' @param object An object of class model_symptome
+#' @param times A list of time points of the simulation period
+#' @param method A string indicating which ode integrator to use. Default is set to 'lsoda'
+#' @return An object of class model_symptome with a dataframe of simulation output
+#'
+setGeneric(name = "ode_simulate",
+           def = function(object, times, method = "lsoda") standardGeneric("ode_simulate"))
+
+setMethod("ode_simulate", signature(object = "model_symptome"),
+  function(object, times, method) {
+    # initial population groups
+    pop_groups <- c(S = object@initial_population$S,
+                    E = object@initial_population$E,
+                    I = object@initial_population$I,
+                    HA = object@initial_population$HA,
+                    R = object@initial_population$R,
+                    CM = object@initial_population$CM)
+    # parameters
+    params <- c(lam = object@parameters$lam,
+                gamma = object@parameters$gamma,
+                omega = object@parameters$omega,
+                ihr = object@parameters$ihr,
+                nu = object@parameters$nu,
+                pdeath = object@parameters$pdeath)
+
+    # ODE system RHS
+    ode_symptome_rhs <- function(t, pop_groups, parameters) {
+      with(
+        as.list(c(pop_groups, params)), {
+          dS <- -lam * S * (I + HA) + omega * R
+          dE <- lam * S * (I + HA) - gamma * E
+          dI <- (1.0 - ihr) * gamma * E - nu.i * I  - pdeath.i * I
+          dHA <- ihr * gamma * E - nu.hsa * HA  - pdeath.hsa * HA
+          dR <- -omega * R + nu.i * I + nu.hsa * HA
+          dCM <-  pdeath.i * I + pdeath.hsa * HA
+          # return the rate of change
+          list(c(dS, dE, dI, dHA, dR, dCM))
+        })
+    }
+    # solving ode
+    output <- ode(y = pop_groups,
+                  times = times,
+                  func = ode_symptome_rhs,
+                  parms = params,
+                  method = method)
+    output <- as.data.frame(output)
+    object@output <- output
+    return(object)
+    })
+
+#' @describeIn model_symptome to plot the outcome of the ode similuation
```

Comment:
I think this would probably work better as a separate plotting function as it could probably be written to take _any_ model type as input.